### DISCUSSION TITLE
marp-cli async rules

### DISCUSSION THREAD

**Discussion by GuillaumeDesforges** [marp-team/marp#466] [Ideas] [answered]
In order to run some rendering, in my case `renderMermaid` from mermaid, one needs to be able to run async rules.

However `markdown-it`'s author is very clear about async rules: they are not and won't be supported.

References:
- https://github.com/markdown-it/markdown-it/blob/master/docs/development.md#i-need-async-rule-how-to-do-it
- https://github.com/markdown-it/markdown-it/issues/248

It states:
> Alternatively, you can render HTML, then parse it to DOM

I've tried to make my own engine:

```js
class PostprocessMarpitEngine extends Marpit {
    constructor(options, postprocess) {
        super(options)
        this.postprocess = postprocess
    }

    withPostprocess(postprocess) {
        this.postprocess = postprocess
        return this
    }

    async render(markdown, env = {}) {
        const { html, css, comments } = super.render(markdown, env)
        if (this.postprocess) {
            return await this.postprocess(markdown, env, html, css, comments)
        }
        return { html, css, comments }
    }
}
```

which I hope to use like so

```js
export default {
    engine: async (constructorOptions) =>
        new PostprocessMarpitEngine(constructorOptions)
            .use(marpMermaidPlugin)
            .withPostprocess(async (markdown, env, html, css, comments) => {
                // do async things
                return { html, css, comments }
            }),
}
```

However the current implementation does not support an async render method.

I'm wondering if changing this line like so would be enough
https://github.com/marp-team/marp-cli/blob/8c7a09bfe8e983ca3bad709a15d42b5cf94ea4ac/src/converter.ts#L174
```diff
-        const ret = engine.render(stripBOM(markdown))
+        const ret = await engine.render(stripBOM(markdown))
```

I don't see any blocker, since the wrapping function is already async
https://github.com/marp-team/marp-cli/blob/8c7a09bfe8e983ca3bad709a15d42b5cf94ea4ac/src/converter.ts#L168C7-L168C15

Would this change be possible?

EDIT: I went ahead and proposed the change.
https://github.com/marp-team/marp-cli/pull/540
I'll try to use it on my own project.

**Comment by yhatt** [comment, marked as answer]
Shipped as [Marp CLI v3.2.1](https://github.com/marp-team/marp-cli/releases/v3.2.1). However, keep in mind returning `Promise` in `Marpit.prototype.render` is still out of the spec of current Marpit framework. _The custom engine with `Promise` will not become working on other Marp/Marpit tools that are strictly following API interface of the framework._

**Reply by GuillaumeDesforges** [reply]
Thanks for the reminder. What are the preferred steps to make it part of the spec?
